In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

import warnings
warnings.filterwarnings('ignore')

In [2]:
anime = pd.read_csv('anime.csv')
rating = pd.read_csv('rating.csv')

#### 애니메이션 데이터 확인

In [3]:
anime.head()

anime_id                              name  \
0     32281                    Kimi no Na wa.   
1      5114  Fullmetal Alchemist: Brotherhood   
2     28977                          Gintama°   
3      9253                       Steins;Gate   
4      9969                     Gintama&#039;   

                                               genre   type episodes  rating  \
0               Drama, Romance, School, Supernatural  Movie        1    9.37   
1  Action, Adventure, Drama, Fantasy, Magic, Mili...     TV       64    9.26   
2  Action, Comedy, Historical, Parody, Samurai, S...     TV       51    9.25   
3                                   Sci-Fi, Thriller     TV       24    9.17   
4  Action, Comedy, Historical, Parody, Samurai, S...     TV       51    9.16   

   members  
0   200630  
1   793665  
2   114262  
3   673572  
4   151266

In [4]:
anime.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 12294 entries, 0 to 12293
Data columns (total 7 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   anime_id  12294 non-null  int64  
 1   name      12294 non-null  object 
 2   genre     12232 non-null  object 
 3   type      12269 non-null  object 
 4   episodes  12294 non-null  object 
 5   rating    12064 non-null  float64
 6   members   12294 non-null  int64  
dtypes: float64(1), int64(2), object(4)
memory usage: 672.5+ KB


In [6]:
anime.columns

Index(['anime_id', 'name', 'genre', 'type', 'episodes', 'rating', 'members'], dtype='object')

In [7]:
anime.type.unique()

array(['Movie', 'TV', 'OVA', 'Special', 'Music', 'ONA', nan], dtype=object)

- **anime 데이터프레임에는 애니메이션의 ID, 이름, 장르와 타입, 에피소드, 사용자 평가 점수 (10점 만점 기준), 시청한 사용자의 수에 대한 정보가 있다.**
- **장르의 경우는 여러 문자열이 콤마를 기준으로 연결되어 있는 것을 확인할 수 있다.**
- **타입의 경우는 음.. 영화냐 드라마냐 음악이냐 하는 것 같다. nan이 있네**

In [8]:
anime['name'] = anime.name.str.replace('&#039;', '')
anime['name'] = anime.name.str.replace('&quot;', '')

- **애니메이션 이름에 특수문자가 html 형태로 섞여있다. 정확한 검색이 가능하도록 제거해준다.**

In [9]:
anime

anime_id                                               name  \
0         32281                                     Kimi no Na wa.   
1          5114                   Fullmetal Alchemist: Brotherhood   
2         28977                                           Gintama°   
3          9253                                        Steins;Gate   
4          9969                                            Gintama   
...         ...                                                ...   
12289      9316       Toushindai My Lover: Minami tai Mecha-Minami   
12290      5543                                        Under World   
12291      5621                     Violence Gekiga David no Hoshi   
12292      6133  Violence Gekiga Shin David no Hoshi: Inma Dens...   
12293     26081                   Yasuji no Pornorama: Yacchimae!!   

                                                   genre   type episodes  \
0                   Drama, Romance, School, Supernatural  Movie        1   
1      Action, Adventure, Drama, Fantasy, Magic, Mili...     TV       64   
2      Action, Comedy, Historical, Parody, Samurai, S...     TV       51   
3                                       Sci-Fi, Thriller     TV       24   
4      Action, Comedy, Historical, Parody, Samurai, S...     TV       51   
...                                                  ...    ...      ...   
12289                                             Hentai    OVA        1   
12290                                             Hentai    OVA        1   
12291                                             Hentai    OVA        4   
12292                                             Hentai    OVA        1   
12293                                             Hentai  Movie        1   

       rating  members  
0        9.37   200630  
1        9.26   793665  
2        9.25   114262  
3        9.17   673572  
4        9.16   151266  
...       ...      ...  
12289    4.15      211  
12290    4.28      183  
12291    4.88      219  
12292    4.98      175  
12293    5.46      142  

[12294 rows x 7 columns]

#### 평가 데이터 확인

In [10]:
rating.head()

user_id  anime_id  rating
0        1        20      -1
1        1        24      -1
2        1        79      -1
3        1       226      -1
4        1       241      -1

In [11]:
rating.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7813737 entries, 0 to 7813736
Data columns (total 3 columns):
 #   Column    Dtype
---  ------    -----
 0   user_id   int64
 1   anime_id  int64
 2   rating    int64
dtypes: int64(3)
memory usage: 178.8 MB


- **rating 컬럼에서 -1은 사용하자 평가하지 않았다는 의미이다. 그러므로 rating 컬럼의 값이 (-1)인 경우 삭제**

In [12]:
user_rating = rating[rating.rating != -1]
user_rating.reset_index(drop = True, inplace = True)
user_rating

user_id  anime_id  rating
0              1      8074      10
1              1     11617      10
2              1     11757      10
3              1     15451      10
4              2     11771      10
...          ...       ...     ...
6337236    73515     16512       7
6337237    73515     17187       9
6337238    73515     22145      10
6337239    73516       790       9
6337240    73516      8074       9

[6337241 rows x 3 columns]

In [13]:
anime.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 12294 entries, 0 to 12293
Data columns (total 7 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   anime_id  12294 non-null  int64  
 1   name      12294 non-null  object 
 2   genre     12232 non-null  object 
 3   type      12269 non-null  object 
 4   episodes  12294 non-null  object 
 5   rating    12064 non-null  float64
 6   members   12294 non-null  int64  
dtypes: float64(1), int64(2), object(4)
memory usage: 672.5+ KB


In [14]:
user_rating.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6337241 entries, 0 to 6337240
Data columns (total 3 columns):
 #   Column    Dtype
---  ------    -----
 0   user_id   int64
 1   anime_id  int64
 2   rating    int64
dtypes: int64(3)
memory usage: 145.0 MB


- **애니메 테이블에 있는 애니메이션 중 유저 평가 테이블에 있는 애니메이션만 남기기**

In [15]:
rated_animeID_list = user_rating['anime_id'].unique()
anime = anime[anime.anime_id.isin(rated_animeID_list)]
anime

anime_id                                               name  \
0         32281                                     Kimi no Na wa.   
1          5114                   Fullmetal Alchemist: Brotherhood   
2         28977                                           Gintama°   
3          9253                                        Steins;Gate   
4          9969                                            Gintama   
...         ...                                                ...   
12288      5541                                   The Satisfaction   
12289      9316       Toushindai My Lover: Minami tai Mecha-Minami   
12290      5543                                        Under World   
12291      5621                     Violence Gekiga David no Hoshi   
12292      6133  Violence Gekiga Shin David no Hoshi: Inma Dens...   

                                                   genre   type episodes  \
0                   Drama, Romance, School, Supernatural  Movie        1   
1      Action, Adventure, Drama, Fantasy, Magic, Mili...     TV       64   
2      Action, Comedy, Historical, Parody, Samurai, S...     TV       51   
3                                       Sci-Fi, Thriller     TV       24   
4      Action, Comedy, Historical, Parody, Samurai, S...     TV       51   
...                                                  ...    ...      ...   
12288                                             Hentai    OVA        1   
12289                                             Hentai    OVA        1   
12290                                             Hentai    OVA        1   
12291                                             Hentai    OVA        4   
12292                                             Hentai    OVA        1   

       rating  members  
0        9.37   200630  
1        9.26   793665  
2        9.25   114262  
3        9.17   673572  
4        9.16   151266  
...       ...      ...  
12288    4.37      166  
12289    4.15      211  
12290    4.28      183  
12291    4.88      219  
12292    4.98      175  

[9926 rows x 7 columns]

In [16]:
anime.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 9926 entries, 0 to 12292
Data columns (total 7 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   anime_id  9926 non-null   int64  
 1   name      9926 non-null   object 
 2   genre     9894 non-null   object 
 3   type      9925 non-null   object 
 4   episodes  9926 non-null   object 
 5   rating    9924 non-null   float64
 6   members   9926 non-null   int64  
dtypes: float64(1), int64(2), object(4)
memory usage: 620.4+ KB


- **사용자가 평가한 애니메이션만 남겨두니 애니메이션 데이터가 약 12,294개에서 9,926개로 데이터가 훅 줄었다.**

#### 애니메이션 추천 시스템 만들기

- **애니메이션 타입 별로 태그처리를 한다. 아까 전에 nan이 있었기 때문에 nan을 제외하고 태그처리하기 위한 타입 리스트를 만든다.**
- **이어서 타입 리스트에 있는 값들을 이름으로 하는 컬럼을 만들어줄거다.그리고 anime\[type\] 값에 맞춰서 O라는 값을 채워줄 것이다.**
- **그리고 type 컬럼은 필요없으니까 버리기~!**

In [17]:
type_list = anime.type.unique()[:-1]
type_list

array(['Movie', 'TV', 'OVA', 'Special', 'Music', 'ONA'], dtype=object)

In [18]:
anime[type_list] = "-"

for t in type_list:
    anime[t] = np.where(anime['type'] == t, '○', '-')

anime = anime.drop(columns = ['type', 'episodes', 'members'])
anime

anime_id                                               name  \
0         32281                                     Kimi no Na wa.   
1          5114                   Fullmetal Alchemist: Brotherhood   
2         28977                                           Gintama°   
3          9253                                        Steins;Gate   
4          9969                                            Gintama   
...         ...                                                ...   
12288      5541                                   The Satisfaction   
12289      9316       Toushindai My Lover: Minami tai Mecha-Minami   
12290      5543                                        Under World   
12291      5621                     Violence Gekiga David no Hoshi   
12292      6133  Violence Gekiga Shin David no Hoshi: Inma Dens...   

                                                   genre  rating Movie TV OVA  \
0                   Drama, Romance, School, Supernatural    9.37     ○  -   -   
1      Action, Adventure, Drama, Fantasy, Magic, Mili...    9.26     -  ○   -   
2      Action, Comedy, Historical, Parody, Samurai, S...    9.25     -  ○   -   
3                                       Sci-Fi, Thriller    9.17     -  ○   -   
4      Action, Comedy, Historical, Parody, Samurai, S...    9.16     -  ○   -   
...                                                  ...     ...   ... ..  ..   
12288                                             Hentai    4.37     -  -   ○   
12289                                             Hentai    4.15     -  -   ○   
12290                                             Hentai    4.28     -  -   ○   
12291                                             Hentai    4.88     -  -   ○   
12292                                             Hentai    4.98     -  -   ○   

      Special Music ONA  
0           -     -   -  
1           -     -   -  
2           -     -   -  
3           -     -   -  
4           -     -   -  
...       ...   ...  ..  
12288       -     -   -  
12289       -     -   -  
12290       -     -   -  
12291       -     -   -  
12292       -     -   -  

[9926 rows x 10 columns]

- **장르에 대해서도 똑같이 해준다!**
- **NaN을 필터링 해주면서 장르를 뽑아내야 하는데, NaN이 float 형태라서, 뽑아낸 장르의 타입이 0.1과 같은 타입이라면 현재 반복을 건너 뛰고 다음 반복으로 넘어가게 한다. 그 이후 뽑아낸 장르를 장르 딕셔너리의 키 값으로 추가한다.**

In [20]:
genre_dict = {}
genre_list = []

for strs in anime.genre.str.split(', '):
    if type(strs) == type(0.1): # Nan 필터링
        continue
    
    for g in strs:
        genre_dict[g] = 0

for key, val in genre_dict.items():
    genre_list.append(key)

type_list = type_list.tolist()
type_list += genre_list
genre_list

['Drama',
 'Romance',
 'School',
 'Supernatural',
 'Action',
 'Adventure',
 'Fantasy',
 'Magic',
 'Military',
 'Shounen',
 'Comedy',
 'Historical',
 'Parody',
 'Samurai',
 'Sci-Fi',
 'Thriller',
 'Sports',
 'Super Power',
 'Space',
 'Slice of Life',
 'Mecha',
 'Music',
 'Mystery',
 'Seinen',
 'Martial Arts',
 'Vampire',
 'Shoujo',
 'Horror',
 'Police',
 'Psychological',
 'Demons',
 'Ecchi',
 'Josei',
 'Shounen Ai',
 'Game',
 'Dementia',
 'Harem',
 'Cars',
 'Kids',
 'Shoujo Ai',
 'Hentai',
 'Yaoi',
 'Yuri']

In [21]:
anime.head(1)

anime_id            name                                 genre  rating  \
0     32281  Kimi no Na wa.  Drama, Romance, School, Supernatural    9.37   

  Movie TV OVA Special Music ONA  
0     ○  -   -       -     -   -

- **장르 리스트를 정리했다. 이제 타입 리스트에 있는 각 타입을 컬럼으로 만들어 채워준 것처럼 장르에 대해서도 똑같이 진핸한다.**
- **np.where의 find 옵션을 써서, 장르 리스트에 있는 g가 해당 데이터 행의 장르 칸에 없다면! -1을 반환한다. 따라서, 데이터 행의 장르 칸에 있으면 anime\[g\]에 O를 넣어주도록 한 것이당.**

In [22]:
for g in genre_list:
    anime[g] = np.where(anime.genre.str.find(g) != -1, '○', '-')

data_anime = anime.drop(['genre'], axis = 1)
data_anime = data_anime.set_index('anime_id')
data_anime

name  rating Movie TV  \
anime_id                                                                       
32281                                        Kimi no Na wa.    9.37     ○  -   
5114                       Fullmetal Alchemist: Brotherhood    9.26     -  ○   
28977                                              Gintama°    9.25     -  ○   
9253                                            Steins;Gate    9.17     -  ○   
9969                                                Gintama    9.16     -  ○   
...                                                     ...     ...   ... ..   
5541                                       The Satisfaction    4.37     -  -   
9316           Toushindai My Lover: Minami tai Mecha-Minami    4.15     -  -   
5543                                            Under World    4.28     -  -   
5621                         Violence Gekiga David no Hoshi    4.88     -  -   
6133      Violence Gekiga Shin David no Hoshi: Inma Dens...    4.98     -  -   

         OVA Special Music ONA Drama Romance  ... Shounen Ai Game Dementia  \
anime_id                                      ...                            
32281      -       -     -   -     ○       ○  ...          -    -        -   
5114       -       -     -   -     ○       -  ...          -    -        -   
28977      -       -     -   -     -       -  ...          -    -        -   
9253       -       -     -   -     -       -  ...          -    -        -   
9969       -       -     -   -     -       -  ...          -    -        -   
...       ..     ...   ...  ..   ...     ...  ...        ...  ...      ...   
5541       ○       -     -   -     -       -  ...          -    -        -   
9316       ○       -     -   -     -       -  ...          -    -        -   
5543       ○       -     -   -     -       -  ...          -    -        -   
5621       ○       -     -   -     -       -  ...          -    -        -   
6133       ○       -     -   -     -       -  ...          -    -        -   

         Harem Cars Kids Shoujo Ai Hentai Yaoi Yuri  
anime_id                                             
32281        -    -    -         -      -    -    -  
5114         -    -    -         -      -    -    -  
28977        -    -    -         -      -    -    -  
9253         -    -    -         -      -    -    -  
9969         -    -    -         -      -    -    -  
...        ...  ...  ...       ...    ...  ...  ...  
5541         -    -    -         -      ○    -    -  
9316         -    -    -         -      ○    -    -  
5543         -    -    -         -      ○    -    -  
5621         -    -    -         -      ○    -    -  
6133         -    -    -         -      ○    -    -  

[9926 rows x 50 columns]

- **필요없는 컬럼인 장르 컬럼을 삭제하고, 테이블 인덱스 값은 anime_id로 정의해줬다.**

In [24]:
mergeDF = pd.merge(user_rating, data_anime['name'], on='anime_id')
mergeDF

user_id  anime_id  rating                                name
0              1      8074      10              Highschool of the Dead
1              3      8074       6              Highschool of the Dead
2              5      8074       2              Highschool of the Dead
3             12      8074       6              Highschool of the Dead
4             14      8074       6              Highschool of the Dead
...          ...       ...     ...                                 ...
6337234    69964     23585       7  Haha wo Tazunete Sanzenri Specials
6337235    69964     33659       6      Fushigi na Somera-chan Special
6337236    72800     30738       4        Gamba: Gamba to Nakama-tachi
6337237    73135      8723       5                    Anime Rakugo Kan
6337238    73188      8749       6                            Forestry

[6337239 rows x 4 columns]

- **사용자 평가 테이블에 유저 아이디 기준으로 영화 제목을 머지해줬다.**

In [25]:
rating_matrix = mergeDF.pivot_table(values = 'rating', index = 'user_id', columns = 'name')
rating_matrix

name     .hack//G.U. Returner  .hack//G.U. Trilogy  \
user_id                                              
1                         NaN                  NaN   
2                         NaN                  NaN   
3                         NaN                  NaN   
5                         NaN                  NaN   
7                         NaN                  NaN   
...                       ...                  ...   
73512                     NaN                  NaN   
73513                     NaN                  NaN   
73514                     NaN                  NaN   
73515                     NaN                  NaN   
73516                     NaN                  NaN   

name     .hack//G.U. Trilogy: Parody Mode  .hack//Gift  .hack//Intermezzo  \
user_id                                                                     
1                                     NaN          NaN                NaN   
2                                     NaN          NaN                NaN   
3                                     NaN          NaN                NaN   
5                                     NaN          NaN                NaN   
7                                     NaN          NaN                NaN   
...                                   ...          ...                ...   
73512                                 NaN          NaN                NaN   
73513                                 NaN          NaN                NaN   
73514                                 NaN          NaN                NaN   
73515                                 NaN          NaN                NaN   
73516                                 NaN          NaN                NaN   

name     .hack//Liminality  .hack//Quantum  \
user_id                                      
1                      NaN             NaN   
2                      NaN             NaN   
3                      NaN             NaN   
5                      NaN             NaN   
7                      NaN             NaN   
...                    ...             ...   
73512                  NaN             NaN   
73513                  NaN             NaN   
73514                  NaN             NaN   
73515                  NaN             NaN   
73516                  NaN             NaN   

name     .hack//Quantum: Sore ike! Bokura no Chimuchimu-chan!!  .hack//Roots  \
user_id                                                                        
1                                                      NaN               NaN   
2                                                      NaN               NaN   
3                                                      NaN               NaN   
5                                                      NaN               NaN   
7                                                      NaN               NaN   
...                                                    ...               ...   
73512                                                  NaN               NaN   
73513                                                  NaN               NaN   
73514                                                  NaN               NaN   
73515                                                  NaN               NaN   
73516                                                  NaN               NaN   

name     .hack//Sign  ...  lilac (bombs Jun Togawa)  makemagic  s.CRY.ed  \
user_id               ...                                                  
1                NaN  ...                       NaN        NaN       NaN   
2                NaN  ...                       NaN        NaN       NaN   
3                NaN  ...                       NaN        NaN       NaN   
5                NaN  ...                       NaN        NaN       NaN   
7                NaN  ...                       NaN        NaN       NaN   
...              ...  ...                       ...        ...       ...   
73512            NaN  ...                      

- **머지한 테이블을 가지고 full matrix 구조를 만들어준다. 그럼 이제 각 유저가 평가한 점수들이 각 애니메이션 별로 테이블이 정리되어 구조가 만들어진다. 전체적으로는 NaN밖에 안보이는데, 각각의 유저들이 평가한 애니메이션에는 유저들의 평가 점수가 들어가 있다.**
- **이제는 NaN 값 때문에 연산 오류가 나는 일을 막기 위해, 모두 0으로 바꾸어준다.**

In [26]:
rating_matrix_fillna = rating_matrix.fillna(0)
rating_matrix_fillna

name     .hack//G.U. Returner  .hack//G.U. Trilogy  \
user_id                                              
1                         0.0                  0.0   
2                         0.0                  0.0   
3                         0.0                  0.0   
5                         0.0                  0.0   
7                         0.0                  0.0   
...                       ...                  ...   
73512                     0.0                  0.0   
73513                     0.0                  0.0   
73514                     0.0                  0.0   
73515                     0.0                  0.0   
73516                     0.0                  0.0   

name     .hack//G.U. Trilogy: Parody Mode  .hack//Gift  .hack//Intermezzo  \
user_id                                                                     
1                                     0.0          0.0                0.0   
2                                     0.0          0.0                0.0   
3                                     0.0          0.0                0.0   
5                                     0.0          0.0                0.0   
7                                     0.0          0.0                0.0   
...                                   ...          ...                ...   
73512                                 0.0          0.0                0.0   
73513                                 0.0          0.0                0.0   
73514                                 0.0          0.0                0.0   
73515                                 0.0          0.0                0.0   
73516                                 0.0          0.0                0.0   

name     .hack//Liminality  .hack//Quantum  \
user_id                                      
1                      0.0             0.0   
2                      0.0             0.0   
3                      0.0             0.0   
5                      0.0             0.0   
7                      0.0             0.0   
...                    ...             ...   
73512                  0.0             0.0   
73513                  0.0             0.0   
73514                  0.0             0.0   
73515                  0.0             0.0   
73516                  0.0             0.0   

name     .hack//Quantum: Sore ike! Bokura no Chimuchimu-chan!!  .hack//Roots  \
user_id                                                                        
1                                                      0.0               0.0   
2                                                      0.0               0.0   
3                                                      0.0               0.0   
5                                                      0.0               0.0   
7                                                      0.0               0.0   
...                                                    ...               ...   
73512                                                  0.0               0.0   
73513                                                  0.0               0.0   
73514                                                  0.0               0.0   
73515                                                  0.0               0.0   
73516                                                  0.0               0.0   

name     .hack//Sign  ...  lilac (bombs Jun Togawa)  makemagic  s.CRY.ed  \
user_id               ...                                                  
1                0.0  ...                       0.0        0.0       0.0   
2                0.0  ...                       0.0        0.0       0.0   
3                0.0  ...                       0.0        0.0       0.0   
5                0.0  ...                       0.0        0.0       0.0   
7                0.0  ...                       0.0        0.0       0.0   
...              ...  ...                       ...        ...       ...   
73512            0.0  ...                      

In [27]:
# 만약 위 코드가 안되면 이렇게 일일이 할 수도 있지만 좋은 뱅법은 아뉘댱.ㅎ 비효율이에요
# for col, item in rating_matix.iteritems():
#     rating_matrix[col] = item.fillna(0)

# rating_matrix

In [28]:
rating_matrix_fillna_np = rating_matrix_fillna.values.T
rating_matrix_fillna_np.shape

(9922, 69600)

- **기존 데이터프레임은 69600 x 9922 인데, 특정 애니메이션을 선택할 시 비슷한 애니메이션을 추천해줄 수 있는 시스템을 만들기 위해서는 위에서 만든 매트릭스의 정보 값의 행열 기준을 반전시켜준다. 이전에는 애니메이션을 컬럼으로, 유저 아이디를 행으로 두었는데, Transpose를 통해 애니메이션을 행으로, 유저 아이디를 컬럼으로 설정하는 것이다.**
- **행열 기준 반전을 통해 69600 x 9922 의 데이터프레임이 9922 x 69600이 된 것을 확인할 수 있다.**

In [29]:
rating_matrix_fillna_np

array([[0., 0., 0., ..., 0., 0., 0.],
       [0., 0., 0., ..., 0., 0., 0.],
       [0., 0., 0., ..., 0., 0., 0.],
       ...,
       [0., 0., 0., ..., 0., 0., 0.],
       [0., 0., 0., ..., 0., 0., 0.],
       [0., 0., 0., ..., 0., 0., 0.]])

#### 추천 시스템 학습
- **TruncatedSVD를 활용해서 학습을 시킨다. full matrix 전체 테이블을 가지고 사용자 기반 협업 필터링을 통해 추천 값을 리스트로 뽑아 보여줄 수도 있지만 현재 테이블이 개인적으로 프로젝트를 하기에는 메모리를 너무 크게 사용한다. 따라서, 행렬 분해 방식을 통해 애니메이션 별로 연관성을 학습하고, 연산할 수 있는 테이블을 통해 값을 뽑아내는 방식을 쓰기 위해 TruncatedSVD를 사용하기로!**

In [30]:
user_count, ani_count = rating_matrix_fillna_np.shape

from sklearn.decomposition import TruncatedSVD
SVD = TruncatedSVD(n_components = 12)
matrix = SVD.fit_transform(rating_matrix_fillna_np)
matrix.shape

(9922, 12)

In [31]:
corr= np.corrcoef(matrix)
corr.shape

(9922, 9922)

- **TruncatedSVD를 통해 뽑아낸 정보를 가지고 피어슨 상관계수를 가져와서 계산한다. 상관계수를 기준으로 특정 애니메이션과 가장 비슷한 애니메이션을 찾아올 수 있도록 해주는 것이다.**

In [32]:
title_list = list(rating_matrix_fillna.columns)
title_list

['.hack//G.U. Returner',
 '.hack//G.U. Trilogy',
 '.hack//G.U. Trilogy: Parody Mode',
 '.hack//Gift',
 '.hack//Intermezzo',
 '.hack//Liminality',
 '.hack//Quantum',
 '.hack//Quantum: Sore ike! Bokura no Chimuchimu-chan!!',
 '.hack//Roots',
 '.hack//Sign',
 '.hack//Tasogare no Udewa Densetsu',
 '.hack//Tasogare no Udewa Densetsu: Offline de Aimashou',
 '.hack//The Movie: Sekai no Mukou ni',
 '.hack//Unison',
 '.hack//Versus: The Thanatos Report',
 '0',
 '001',
 '009 Re:Cyborg',
 '009-1',
 '009-1: R&amp;B',
 '00:08',
 '07-Ghost',
 '1+2=Paradise',
 '100%',
 '100-man-nen Chikyuu no Tabi: Bander Book',
 '1000-nen Joou: Queen Millennia',
 '1001 Nights',
 '11-nin Iru!',
 '11eyes',
 '11eyes Picture Drama',
 '11eyes: Momoiro Genmutan',
 '12-gatsu no Uta',
 '12-sai.',
 '12-sai. 2nd Season',
 '12-sai.: Chicchana Mune no Tokimeki',
 '12-sai.: Chicchana Mune no Tokimeki 2nd Season',
 '15 Bishoujo Hyouryuuki',
 '15 Sonyeon Uju Pyoryugi',
 '1989',
 '20-dai no Heya-hen',
 '2002-nen Harmoni Ondo',
 '20

#### 추천시스템 완성
- **검색할 수 있는 애니메이션 이름을 리스트로 뽑는다. 이제 애니메이션 검색하면 이 리스트에서 애니메이션 매칭하는거다.**
1. **사용자로부터 애니메이션 제목을 입력값으로 받고, 애니메이션 제목 리스트에서 입력값으로 받은 애니메이션 제목의 인덱스를 찾는다.**
2. **인덱스 값을 기준으로 입력받은 애니메이션의 피어슨 상관계수 값을 계산한다.**
3. **피어슨 상관계수 값이 90% 이상인 애니메이션을 최대 10개를 뽑아서 테이블 형태로 보여준다.**

In [33]:
selected_title = input("Enter the Title of Animation : ")

idx = title_list.index(selected_title)
similarity = corr[idx]

idx = 0
idx_dict = {}

for val in similarity:
    if val >= 0.9:
        idx_dict[val] = title_list[idx]
    idx += 1

ani_names = []
limit = 10
idx = 0
for key, val in sorted (idx_dict.items(), reverse = True):
    if limit < idx:
        break
    if idx == 0:
        idx += 1
        continue
    
    ani_names.append(val)
    idx += 1
    
answer_table = data_anime[data_anime.name.isin(ani_names)]
answer_table

Enter the Title of Animation :  Bokurano


name  rating Movie TV  \
anime_id                                                                       
3901                                      Baccano! Specials    8.29     -  -   
239                                              Gankutsuou    8.27     -  ○   
3167                                           Eve no Jikan    8.26     -  -   
2164                                            Dennou Coil    8.19     -  ○   
1379      Kino no Tabi: Nanika wo Suru Tame ni - Life Go...    7.91     ○  -   
4981                                          Casshern Sins    7.61     -  ○   
2596                               Shinreigari: Ghost Hound    7.59     -  ○   
5356                                                 Canaan    7.40     -  ○   
8248                      Soukou Kihei Votoms: Case; Irvine    6.69     -  -   
8249                                          Votoms Finder    6.39     -  -   

         OVA Special Music ONA Drama Romance  ... Shounen Ai Game Dementia  \
anime_id                                      ...                            
3901       -       ○     -   -     -       -  ...          -    -        -   
239        -       -     -   -     ○       -  ...          -    -        -   
3167       -       -     -   ○     -       -  ...          -    -        -   
2164       -       -     -   -     ○       -  ...          -    -        -   
1379       -       -     -   -     ○       -  ...          -    -        -   
4981       -       -     -   -     ○       -  ...          -    -        -   
2596       -       -     -   -     -       -  ...          -    -        -   
5356       -       -     -   -     -       -  ...          -    -        -   
8248       ○       -     -   -     ○       -  ...          -    -        -   
8249       ○       -     -   -     ○       -  ...          -    -        -   

         Harem Cars Kids Shoujo Ai Hentai Yaoi Yuri  
anime_id                                             
3901         -    -    -         -      -    -    -  
239          -    -    -         -      -    -    -  
3167         -    -    -         -      -    -    -  
2164         -    -    -         -      -    -    -  
1379         -    -    -         -      -    -    -  
4981         -    -    -         -      -    -    -  
2596         -    -    -         -      -    -    -  
5356         -    -    -         -      -    -    -  
8248         -    -    -         -      -    -    -  
8249         -    -    -         -      -    -    -  

[10 rows x 50 columns]

In [34]:
selected_title = input("Enter the Title of Animation : ")

idx = title_list.index(selected_title)
similarity = corr[idx]

idx = 0
idx_dict = {}

for val in similarity:
    if val >= 0.9:
        idx_dict[val] = title_list[idx]
    idx += 1

ani_names = []
limit = 10
idx = 0
for key, val in sorted (idx_dict.items(), reverse = True):
    if limit < idx:
        break
    if idx == 0:
        idx += 1
        continue
    
    ani_names.append(val)
    idx += 1
    
answer_table = data_anime[data_anime.name.isin(ani_names)]
answer_table

Enter the Title of Animation :  Howl no Ugoku Shiro


name  rating Movie TV OVA Special Music  \
anime_id                                                                     
199       Sen to Chihiro no Kamikakushi    8.93     ○  -   -       -     -   
164                       Mononoke Hime    8.81     ○  -   -       -     -   
523                    Tonari no Totoro    8.48     ○  -   -       -     -   
572            Kaze no Tani no Nausicaä    8.47     ○  -   -       -     -   
513              Tenkuu no Shiro Laputa    8.38     ○  -   -       -     -   
585                    Mimi wo Sumaseba    8.32     ○  -   -       -     -   
512                  Majo no Takkyuubin    8.27     ○  -   -       -     -   
597                    Neko no Ongaeshi    7.98     ○  -   -       -     -   
2890                Gake no Ue no Ponyo    7.93     ○  -   -       -     -   
1829                         Gedo Senki    7.18     ○  -   -       -     -   

         ONA Drama Romance  ... Shounen Ai Game Dementia Harem Cars Kids  \
anime_id                    ...                                            
199        -     ○       -  ...          -    -        -     -    -    -   
164        -     -       -  ...          -    -        -     -    -    -   
523        -     -       -  ...          -    -        -     -    -    -   
572        -     -       -  ...          -    -        -     -    -    -   
513        -     -       ○  ...          -    -        -     -    -    -   
585        -     ○       ○  ...          -    -        -     -    -    -   
512        -     ○       ○  ...          -    -        -     -    -    -   
597        -     ○       -  ...          -    -        -     -    -    -   
2890       -     -       -  ...          -    -        -     -    -    -   
1829       -     -       -  ...          -    -        -     -    -    -   

         Shoujo Ai Hentai Yaoi Yuri  
anime_id                             
199              -      -    -    -  
164              -      -    -    -  
523              -      -    -    -  
572              -      -    -    -  
513              -      -    -    -  
585              -      -    -    -  
512              -      -    -    -  
597              -      -    -    -  
2890             -      -    -    -  
1829             -      -    -    -  

[10 rows x 50 columns]

In [ ]:
selected_title = input("Enter the Title of Animation : ")

idx = title_list.index(selected_title)
similarity = corr[idx]

idx = 0
idx_dict = {}

for val in similarity:
    if val >= 0.9:
        idx_dict[val] = title_list[idx]
    idx += 1

ani_names = []
limit = 10
idx = 0
for key, val in sorted (idx_dict.items(), reverse = True):
    if limit < idx:
        break
    if idx == 0:
        idx += 1
        continue
    
    ani_names.append(val)
    idx += 1
    
answer_table = data_anime[data_anime.name.isin(ani_names)]
answer_table